includes/Core/DI/DI_Container.php

```diff
@@ -0,0 +1,258 @@
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param mixed  $entry Entry.
+	 */
+	public function offsetSet( $id, $entry ) {
+		if ( is_callable( $entry ) ) {
+			$this->set_service( $id, $entry );
+		} else {
+			$this->set_value( $id, $entry );
+		}
+	}
+
+	/**
+	 * Removes the entry.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry to remove.
+	 */
+	public function offsetUnset( $id ) {
+		unset( $this->definitions[ $id ] );
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 * @param array  $entry Entry definition.
+	 * @return bool TRUE if the entry is added, otherwise FALSE.
+	 */
+	protected function set( $id, array $entry ) {
+		if ( ! empty( $this->definitions[ $id ]['is_protected'] ) ) {
+			return false;
+		}
+
+		$this->definitions[ $id ] = $entry;
+
+		return true;
+	}
+
+	/**
+	 * Sets the entry definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Entry name.
+	 * @param mixed  $value Entry value.
+	 * @return bool TRUE if the service is added, otherwise FALSE.
+	 */
+	public function set_value( $id, $value ) {
+		return $this->set(
+			$id,
+			array(
+				'entry' => $value,
+			)
+		);
+	}
+
+	/**
+	 * Sets the service definition.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string   $id Service name.
+	 * @param callable $create_func Service creator.
+	 * @return bool TRUE if the service is added, otherwise FALSE.
+	 */
+	public function set_service( $id, $create_func ) {
+		return $this->set(
+			$id,
+			array(
+				'is_service' => true,
+				'entry'      => $create_func,
+			)
+		);
+	}
+
+	/**
+	 * Sets the entry to be factory.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string   $id Identifier of the entry.
+	 * @param callable $factory_func Factory function.
+	 * @return bool TRUE if the factory is added, otherwise FALSE.
+	 */
+	public function set_factory( $id, $factory_func ) {
+		return $this->set(
+			$id,
+			array(
+				'is_service' => true,
+				'is_factory' => true,
+				'entry'      => $factory_func,
+			)
+		);
+	}
+
+	/**
+	 * Sets the entry to be protected.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $id Identifier of the entry.
+	 */
+	public function set_is_protected( $id ) {
+		if ( $this->has( $id ) ) {
+			$this->definitions[ $id ]['is_protected'] = true;
+		}
+	}
+
+	/**
+	 * Sets services.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $definitions Service definitions.
+	 */
+	public function set_services( array $definitions ) {
+		foreach ( $definitions as $service_name => $service_class ) {
+			$service_class_key = sprintf(
+				'%s_CLASS',
+				strtoupper( $service_name )
+			);
+
+			$this->set_value( $service_class_key, $service_class );
+			$this->set_service(
+				$service_name,
+				function( $di ) use ( $service_class_key ) {
+					return new $di[ $service_class_key ]();
+				}
+			);
+		}
```

Comment:
It's weird to me that `set_services` works differently than `set_service`. Related to my other point, why not instead change `set_service` to accept either a class name or create function? IMO the logic should be bundled there and not only work if you call `set_services`. What if you want to set only one service where you wanna pass a class name? It's rather unintuitive that you have to call `set_services` in that case and can't use `set_service` in the same way.

Reply:
Yep, updated it to allow passing class names into the `set_service` method.